src/createStreamMiddleware.ts

```diff
@@ -0,0 +1,96 @@
+import SafeEventEmitter from '@metamask/safe-event-emitter';
+import { Duplex } from 'readable-stream';
+import {
+  JsonRpcEngineNextCallback,
+  JsonRpcEngineEndCallback,
+  JsonRpcNotification,
+  JsonRpcMiddleware,
+  JsonRpcRequest,
+  PendingJsonRpcResponse,
+} from 'json-rpc-engine';
+
+interface IdMapValue {
+  req: JsonRpcRequest<unknown>;
+  res: PendingJsonRpcResponse<unknown>;
+  next: JsonRpcEngineNextCallback;
+  end: JsonRpcEngineEndCallback;
+}
+
+interface IdMap {
+  [requestId: string]: IdMapValue;
+}
+
+/**
+ * Creates a JsonRpcEngine middleware with an associated Duplex stream and
+ * EventEmitter. The middleware, and by extension stream, assume that middleware
+ * parameters are properly formatted. No runtime type checking or validation is
+ * performed.
+ * 
+ * @returns The event emitter, middleware, and stream.
+ */
+export default function createStreamMiddleware() {
+  const idMap: IdMap = {};
+  const stream = new Duplex({
+    objectMode: true,
+    read: readNoop,
+    write: processMessage,
+  });
+
+  const events = new SafeEventEmitter();
+
+  const middleware: JsonRpcMiddleware<unknown, unknown> = (
+    req,
+    res,
+    next,
+    end
```

Comment:
```suggestion
    end,
```
lint error - comma after the last parameter